java/google/registry/rde/imports/RdeHostLinkAction.java

```diff
@@ -0,0 +1,160 @@
+    Optional<InternetDomainName> tld = findTldForName(hostName);
+    // out of zone hosts cannot be linked
+    if(!tld.isPresent()) {
+      return null;
+    }
+    // This is a subordinate host
+    String domainName = Joiner.on('.').join(Iterables.skip(
+        hostName.parts(), hostName.parts().size() - (tld.get().parts().size() + 1)));
+    DomainResource superordinateDomain = loadByForeignKey(DomainResource.class, domainName, now);
+    // Hosts can't be linked if domains import hasn't been run
+    checkState(superordinateDomain != null,
+      "Superordinate domain not found: %s",
+      domainName);
+    return superordinateDomain;
+  }
+
+  /** Mapper to link hosts from an escrow file to their superordinate domains. */
+  public static class RdeHostPostImportMapper extends Mapper<JaxbFragment<XjcRdeHostElement>, Void, Void> {
+
+    private static final long serialVersionUID = -2898753709127134419L;
+
+    @Override
+    public void map(JaxbFragment<XjcRdeHostElement> fragment) {
+      // Record number of attempted map operations
+      getContext().incrementCounter("post-import hosts read");
+      final XjcRdeHost xjcHost = fragment.getInstance().getValue();
+      logger.infofmt("Attempting to link superordinate domain for host %s", xjcHost.getName());
+      try {
+        InternetDomainName hostName = InternetDomainName.from(xjcHost.getName());
+        DomainResource superordinateDomain = lookupSuperordinateDomain(hostName, DateTime.now());
+        // if suporordinateDomain is null, this is an out of zone host and can't be linked
+        if (superordinateDomain == null) {
+          getContext().incrementCounter("post-import hosts out of zone");
+          logger.infofmt("Host %s is out of zone", xjcHost.getName());
+          return;
+        }
+        // at this point, the host is definitely in zone and should be linked
+        getContext().incrementCounter("post-import hosts in zone");
+        final Key<DomainResource> superordinateDomainKey = Key.create(superordinateDomain);
+        ofy().transact(new VoidWork() {
+          @Override
+          public void vrun() {
+            HostResource host =
+                ofy().load().now(Key.create(HostResource.class, xjcHost.getRoid()));
+            ofy().save()
+                .entity(host.asBuilder().setSuperordinateDomain(superordinateDomainKey).build());
+          }
+        });
+        logger.infofmt(
+            "Successfully linked host %s to superordinate domain %s",
+            xjcHost.getName(),
+            superordinateDomain.getFullyQualifiedDomainName());
+        // Record number of hosts successfully linked
+        getContext().incrementCounter("post-import hosts linked");
+      } catch (Exception e) {
+        // Record the number of hosts with unexpected errors
+        getContext().incrementCounter("post-import host errors");
+        throw new HostLinkException(xjcHost.getName(), xjcHost.toString(), e);
```

Comment:
This will cause the entire mapreduce to abort and terminate for a single failing host, rather than continuing to process all of the ones that can be processed successfully.  Is that intended?  Note that other hosts will probably already have been saved by this point, so all operations better be idempotent.